argilla/src/argilla/datasets/_export/_hub.py

```diff
@@ -0,0 +1,188 @@
+    from argilla import Dataset
+    from argilla import Argilla
+    from argilla import Workspace
+    from datasets import Dataset as HFDataset
+
+
+class HubImportExportMixin:
+    _DEFAULT_DATASET_REPO_DIR = ".argilla"
+    _DEFAULT_CONFIGURATION_FILES = ["settings.json", "dataset.json"]
+
+    def to_hub(
+        self: "Dataset",
+        repo_id: str,
+        generate_card: Optional[bool] = True,
+        *args,
+        **kwargs,
+    ) -> None:
+        """Pushes the `Dataset` to the Hugging Face Hub. If the dataset has been previously pushed to the
+        Hugging Face Hub, it will be updated instead of creating a new dataset repo.
+
+        Parameters:
+            repo_id: the ID of the Hugging Face Hub repo to push the `Dataset` to.
+            generate_card: whether to generate a dataset card for the `Dataset` in the Hugging Face Hub. Defaults
+                to `True`.
+            *args: the args to pass to `datasets.Dataset.push_to_hub`.
+            **kwargs: the kwargs to pass to `datasets.Dataset.push_to_hub`.
+        """
+
+        from huggingface_hub import HfApi, DatasetCardData
+
+        from argilla.datasets._export.card import (
+            ArgillaDatasetCard,
+            size_categories_parser,
+        )
+
+        hf_api = HfApi(token=kwargs.get("token"))
+
+        hfds = self.records(with_vectors=True, with_responses=True, with_suggestions=True).to_datasets()
+        hfds.push_to_hub(repo_id, *args, **kwargs)
+
+        with TemporaryDirectory() as tmpdirname:
+            config_dir = os.path.join(tmpdirname, self._DEFAULT_DATASET_REPO_DIR)
+            os.makedirs(config_dir)
+            self.to_disk(path=config_dir, with_records=False)
+
+            if generate_card:
+                sample_argilla_record = next(iter(self.records(with_suggestions=True, with_responses=True)))
+                sample_huggingface_record = hfds[0]
+                card = ArgillaDatasetCard.from_template(
+                    card_data=DatasetCardData(
+                        size_categories=size_categories_parser(len(hfds)),
+                        tags=["rlfh", "argilla", "human-feedback"],
+                    ),
+                    repo_id=repo_id,
+                    argilla_fields=self.settings.fields,
+                    argilla_questions=self.settings.questions,
+                    argilla_guidelines=self.settings.guidelines or None,
+                    argilla_vectors_settings=self.settings.vectors or None,
+                    argilla_metadata_properties=self.settings.metadata,
+                    argilla_record=sample_argilla_record.to_dict(),
+                    huggingface_record=sample_huggingface_record,
+                )
+                card.save(filepath=os.path.join(tmpdirname, "README.md"))
+
+            hf_api.upload_folder(
+                folder_path=tmpdirname,
+                repo_id=repo_id,
+                repo_type="dataset",
+                token=kwargs.get("token"),
```

Comment:
I've noticed that it's a common pattern. I think it's to deal with multiple token role assignments, for example, across organisations. 

@Wauplin You recommended making token consistent, but would you also recommend exposing it as a param like this?

Reply:
There are two topics here:
1. Passing `token=kwargs.get("token")` in `hf_api.upload_folder` is indeed not needed since `hf_api` is instantiated as `hf_api = HfApi(token=kwargs.get("token"))`.
2. Is it necessary to expose `token` as a parameter? => this depends on how Argilla is used. In general the preferred way to authenticate is to set the `HF_TOKEN` environment variable or to run `huggingface-cli login` in a terminal which will save your token in a file on the machine. This is the preferred way since it insures users won't paste their token in the code directly. _However_, it only works if the script/tool perform all actions as the same user. Now if the tool is running on a server and multiple users can access it and perform actions with their own token, then it's useful to have a `token` parameter. For example, a lot of Gradio Spaces have a "Sign in with Hugging Face" button to retrieve a user token. In the process, the Space might need to save an artifact under the user namespace on HF (say a dataset) => it has to use the `token` parameter.